File: internal/rules/rules.go
package rules

import (
	"sort"
	"strings"

	"github.com/jsonbored/nightward/internal/inventory"
)

type Rule struct {
	ID              string              `json:"id"`
	Category        string              `json:"category"`
	DefaultSeverity inventory.RiskLevel `json:"default_severity"`
	Title           string              `json:"title"`
	Description     string              `json:"description"`
	Recommendation  string              `json:"recommended_action"`
	DocsURL         string              `json:"docs_url,omitempty"`
}

func List() []Rule {
	out := []Rule{
		{ID: "mcp_server_review", Category: "mcp", DefaultSeverity: inventory.RiskInfo, Title: "Review MCP server", Description: "Every MCP server is an executable trust boundary that should be reviewed before syncing.", Recommendation: "Confirm source, permissions, and local assumptions before backing up config.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/remediation.md"},
		{ID: "mcp_unpinned_package", Category: "mcp", DefaultSeverity: inventory.RiskHigh, Title: "Unpinned package executor", Description: "An MCP server invokes a package executor without an obvious pinned package version.", Recommendation: "Pin package versions or replace the executor with a reviewed local binary.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/remediation.md"},
		{ID: "mcp_secret_env", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Sensitive environment key", Description: "An MCP server references a sensitive environment key.", Recommendation: "Keep secret values outside portable config and document required env keys.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/privacy-model.md"},
		{ID: "mcp_secret_header", Category: "mcp", DefaultSeverity: inventory.RiskHigh, Title: "Sensitive header key", Description: "A URL-shaped MCP server references a sensitive header key.", Recommendation: "Externalize inline header values and keep only reviewed env references in config.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/privacy-model.md"},
		{ID: "mcp_shell_command", Category: "mcp", DefaultSeverity: inventory.RiskHigh, Title: "Shell wrapper", Description: "An MCP server executes through a shell wrapper.", Recommendation: "Prefer direct executable invocation when the wrapper is a simple passthrough.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/remediation.md"},
		{ID: "mcp_broad_filesystem", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Broad filesystem access", Description: "An MCP server appears to request broad filesystem access.", Recommendation: "Replace broad mounts with explicit reviewed paths when possible.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/remediation.md"},
		{ID: "mcp_local_endpoint", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Local endpoint", Description: "A URL-shaped MCP server points at loopback, localhost, or private network state.", Recommendation: "Document the local service dependency and avoid assuming it exists on every machine.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/privacy-model.md"},
		{ID: "mcp_local_token_path", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Local token path", Description: "An MCP server argument appears to reference local token or credential material.", Recommendation: "Keep token paths out of portable config and recreate them per machine.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/privacy-model.md"},
		{ID: "mcp_symlink_config", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Symlinked MCP config", Description: "An MCP config path resolves through a symlink, which can hide the real source being synced or reviewed.", Recommendation: "Review the resolved target and keep machine-local symlinks out of portable dotfiles.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/privacy-model.md"},
		{ID: "mcp_unknown_command", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "Unknown MCP server shape", Description: "An MCP server lacks both an executable command and URL endpoint.", Recommendation: "Review the config manually and update Nightward fixtures if this is a supported shape.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/adapters.md"},
		{ID: "mcp_parse_failed", Category: "mcp", DefaultSeverity: inventory.RiskMedium, Title: "MCP parse failure", Description: "Nightward could not parse an MCP config file.", Recommendation: "Fix the config syntax, then rerun scan and policy checks.", DocsURL: "https://github.com/JSONbored/nightward/blob/main/docs/adapters.md"},
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func Find(idOrPrefix string) (Rule, bool) {
	if idOrPrefix == "" {
		return Rule{}, false
	}
	for _, rule := range List() {
		if rule.ID == idOrPrefix {
			return rule, true
		}
	}
	var matched []Rule
	for _, rule := range List() {
		if strings.HasPrefix(rule.ID, idOrPrefix) {
			matched = append(matched, rule)
		}
	}
	if len(matched) == 1 {
		return matched[0], true
	}
	return Rule{}, false
}

File: crates/nightward-core/src/rules.rs
use crate::{FixKind, RiskLevel};
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct Rule {
    pub id: &'static str,
    pub severity: RiskLevel,
    pub fix_kind: FixKind,
    pub title: &'static str,
    pub docs_url: &'static str,
}

pub fn all_rules() -> Vec<Rule> {
    vec![
        Rule {
            id: "mcp_secret_env",
            severity: RiskLevel::Critical,
            fix_kind: FixKind::ExternalizeSecret,
            title: "MCP server stores a sensitive environment variable inline",
            docs_url: "https://nightward.aethereal.dev/guide/remediation",
        },
        Rule {
            id: "mcp_secret_header",
            severity: RiskLevel::Critical,
            fix_kind: FixKind::ExternalizeSecret,
            title: "MCP server stores a sensitive header inline",
            docs_url: "https://nightward.aethereal.dev/guide/remediation",
        },
        Rule {
            id: "mcp_unpinned_package",
            severity: RiskLevel::High,
            fix_kind: FixKind::PinPackage,
            title: "MCP server runs a package executor without an obvious pin",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_shell_wrapper",
            severity: RiskLevel::High,
            fix_kind: FixKind::ReplaceShellWrapper,
            title: "MCP server runs through a shell wrapper",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_local_endpoint",
            severity: RiskLevel::Medium,
            fix_kind: FixKind::ManualReview,
            title: "MCP server references a machine-local endpoint",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_broad_filesystem",
            severity: RiskLevel::Medium,
            fix_kind: FixKind::NarrowFilesystem,
            title: "MCP server can access a broad filesystem path",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_local_token_path",
            severity: RiskLevel::High,
            fix_kind: FixKind::ManualReview,
            title: "MCP server references a local credential path",
            docs_url: "https://nightward.aethereal.dev/guide/privacy-model",
        },
        Rule {
            id: "mcp_docker_socket",
            severity: RiskLevel::High,
            fix_kind: FixKind::ManualReview,
            title: "MCP server can control Docker or container host state",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_typosquat_package",
            severity: RiskLevel::Medium,
            fix_kind: FixKind::ManualReview,
            title: "MCP server package resembles a trusted namespace",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_untrusted_package_source",
            severity: RiskLevel::Medium,
            fix_kind: FixKind::ManualReview,
            title: "MCP server launches a remote package or script source",
            docs_url: "https://nightward.aethereal.dev/guide/mcp-security",
        },
        Rule {
            id: "mcp_server_review",
            severity: RiskLevel::Info,
            fix_kind: FixKind::ManualReview,
            title: "MCP server should be reviewed",
            docs_url: "https://nightward.aethereal.dev/reference/rules",
        },
        Rule {
            id: "mcp_unknown_command",
            severity: RiskLevel::Info,
            fix_kind: FixKind::ManualReview,
            title: "MCP server has an unsupported command shape",
            docs_url: "https://nightward.aethereal.dev/reference/rules",
        },
        Rule {
            id: "config_parse_failed",
            severity: RiskLevel::Medium,
            fix_kind: FixKind::ManualReview,
            title: "Nightward could not parse a config file",
            docs_url: "https://nightward.aethereal.dev/use/troubleshooting",
        },
        Rule {
            id: "config_symlink",
            severity: RiskLevel::Info,
            fix_kind: FixKind::ManualReview,
            title: "Config file is a symbolic link",
            docs_url: "https://nightward.aethereal.dev/guide/privacy-model",
        },
        Rule {
            id: "config_stale",
            severity: RiskLevel::Low,
            fix_kind: FixKind::ManualReview,
            title: "Config file has not changed in over 180 days",
            docs_url: "https://nightward.aethereal.dev/guide/privacy-model",
        },
    ]
}

pub fn explain_rule(id: &str) -> Option<Rule> {
    all_rules().into_iter().find(|rule| rule.id == id)
}
